sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlus.scala

```diff
@@ -141,9 +141,11 @@ case class HyperLogLogPlusPlus(
 
 object HyperLogLogPlusPlus {
   def validateDoubleLiteral(exp: Expression): Double = exp match {
-    case Literal(d: Double, DoubleType) => d
-    case Literal(dec: Decimal, _) => dec.toDouble
+    case e if e.foldable && e.dataType == DoubleType =>
+      e.eval().asInstanceOf[Double]
+    case e if e.foldable && e.dataType.isInstanceOf[DecimalType] =>
+      e.eval().asInstanceOf[Decimal].toDouble
```

Comment:
Ah, nice catch! Is this a last place having the same issue?

Reply:
I believed it is but I just find a new at `FirstLast`.

Reply:
https://github.com/apache/spark/blob/3a299aa6480ac22501512cd0310d31a441d7dfdc/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/First.scala#L124-L129

Reply:
Ah, ok. Could you fix that, too, in a separate PR?